# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [35]:
# import libraries
import pandas as pd
import numpy as np

# read csv
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv') 
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [36]:
# Understanding rows and columns
df.shape

(4008, 11)

In [37]:
# checking type of each column
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [38]:

# checking details of each column for missing info
df.info

<bound method DataFrame.info of      Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 7

__Recommendation:__ Remove rows or columns with Nan

In [39]:
# Removing rows with Nan

df_clean_rows = df.dropna()
df_clean_rows

Customer          ST GENDER             Education  \
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
5                 825629.78%  62902.0                  69.0   
...                      ...      ...                   ...   
1066              305955.03%  38644.0                  78.0   
1067             2031499.76%  63209.0                 102.0   
1068              323912.47%  16061.0                  88.0   
1069              462680.11%  79487.0                 114.0   
1070              899704.02%  54230.0                 112.0   

     Number of Open Complaints     Policy Type  Vehicle Class  \
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
5                       1/0/00   Personal Auto   Two-Door Car   
...                        ...             ...            ...   
1066                    1/1/00   Personal Auto  Four-Door Car   
1067                    1/2/00   Personal Auto            SUV   
1068                    1/0/00   Personal Auto  Four-Door Car   
1069                    1/0/00    Special Auto            SUV   
1070                    1/0/00   Personal Auto   Two-Door Car   

      Total Claim Amount  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
5             159.383042  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[952 rows x 11 columns]

In [40]:
# Removing columns with Nan

df_clean_col = df.dropna(axis=1)
df_clean_col 

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[4008 rows x 0 columns]

In [41]:
# Identifying the number of unique values for each column 
df.nunique() 

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [42]:
# Columns with type object and count of unique categorical

cat_cols = df.select_dtypes("object").nunique().sort_values(ascending=False)
cat_cols

Customer                     1071
Customer Lifetime Value      1027
ST                              8
Education                       6
Number of Open Complaints       6
Vehicle Class                   6
GENDER                          5
Policy Type                     3
dtype: int64

__Insights to data__

- Columns with a small number of unique values relative to total row count are known as categorical (e.g. gender, policy type, state).
- Columns with numeric data type and high cardinality are considered  continuous/numerical (e.g. income, age, amounts).

If a column is encoded with numbers but has low unique values (e.g. 0/1 for gender), I would treat as categorical.


In [43]:
# Calculation statistics for each numeric column

num_cols = df.select_dtypes(include=['float64']).columns.tolist()

for col in num_cols:
    mean = df[col].mean()
    median = df[col].median()
    mode = df[col].mode()
    variance = df[col].var()
    std_dev = df[col].std()
    min_value = df[col].min()
    max_value = df[col].max()
    range_value = max_value - min_value
    quantiles = df[col].quantile([0.25, 0.5, 0.75])
    
    # Print the results
    print(f"Statistics for {col}:")
    print(f"  Mean: {mean}")
    print(f"  Median: {median}")
    print(f"  Mode: {mode}")
    print(f"  Variance: {variance}")
    print(f"  Standard Deviation: {std_dev}")
    print(f"  Minimum: {min_value}")
    print(f"  Maximum: {max_value}")
    print(f"  Range: {range_value}")
    print(f"  Quantiles (25%, 50%, 75%): {quantiles}")
    print()

Statistics for Income:
  Mean: 39295.70121381886
  Median: 36234.0
  Mode: 0    0.0
Name: Income, dtype: float64
  Variance: 928385985.3891485
  Standard Deviation: 30469.42706040185
  Minimum: 0.0
  Maximum: 99960.0
  Range: 99960.0
  Quantiles (25%, 50%, 75%): 0.25    14072.0
0.50    36234.0
0.75    64631.0
Name: Income, dtype: float64

Statistics for Monthly Premium Auto:
  Mean: 193.234360410831
  Median: 83.0
  Mode: 0    65.0
Name: Monthly Premium Auto, dtype: float64
  Variance: 2563810.596425738
  Standard Deviation: 1601.190368577621
  Minimum: 61.0
  Maximum: 35354.0
  Range: 35293.0
  Quantiles (25%, 50%, 75%): 0.25     68.0
0.50     83.0
0.75    109.5
Name: Monthly Premium Auto, dtype: float64

Statistics for Total Claim Amount:
  Mean: 404.98690940896364
  Median: 354.729129
  Mode: 0    321.6
Name: Total Claim Amount, dtype: float64
  Variance: 85864.97505997638
  Standard Deviation: 293.0272599264041
  Minimum: 0.382107
  Maximum: 2893.239678
  Range: 2892.857571
  Quant

__Observations:__

- Income has a wide range and high variance, indicating significant variability among individuals' incomes.
- Monthly Premium Auto shows a lower range, suggesting most values cluster within a smaller interval.
- Total Claim Amount also reflects variability, but has a lower mean compared to Income, indicating claims are generally less than income values.

In [44]:
df[cat_col].describe()

ST Education Number of Open Complaints  Vehicle Class GENDER  \
count     1071      1071                      1071           1071    954   
unique       8         6                         6              6      5   
top     Oregon  Bachelor                    1/0/00  Four-Door Car      F   
freq       320       324                       830            576    457   

          Policy Type  
count            1071  
unique              3  
top     Personal Auto  
freq              780

__Summary of Conclusions:__

- State (ST):
There are 8 unique states represented in the dataset. The top State Oregon has the highest count of 320 customers 

- Education:
There are 6 unique education levels present. The most common education level is "Bachelor," with a frequency of 32

- Number of Open Complaints:
There are 6 unique values for the number of open complaints. The most frequent value is 1 open complaint, with 830 occurrences.

- Vehicle Class:
There are 6 unique vehicle classes. Four-Door Car is the most common, with 576 customers.

- Gender:
There are 5 unique gender categories, with a total of 954 entries. Female (F) is the most represented gender, with 457 customers.

- Policy Type:
There are 3 unique policy types. Personal Auto is the most common policy type, with 780 instances.


## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [45]:

# Counting the occurrences of each location
location_counts = df_clean_rows['ST'].value_counts()

# Sorting the counts in ascending order
sorted_locations = location_counts.sort_values()

# The top 5 less common locations
less_common_locations = sorted_locations.head(5)

print(less_common_locations)

ST
AZ             14
WA             29
Washington     77
Nevada         89
Cali          119
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [48]:
# A count of the number of policies sold
policy_counts = df['Policy Type'].value_counts()

# The policy type with the highest number of policies sold
most_common_policy = policy_counts.idxmax() 
max_count = policy_counts.max()  

print("Total number of policies sold for each type:")
print(policy_counts)
print("\nPolicy type with the highest number of policies sold:")
print(most_common_policy, "with", max_count, "policies sold.")

Total number of policies sold for each type:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Policy type with the highest number of policies sold:
Personal Auto with 780 policies sold.


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [52]:

# Creating a dataframe for Personal Auto and Corporate Auto policies
personal_auto_df = df.loc[df['Policy Type'] == 'Personal Auto']
corporate_auto_df = df.loc[df['Policy Type'] == 'Corporate Auto']

# Calculating the average income for each policy type
average_income_personal = personal_auto_df['Income'].mean()
average_income_corporate = corporate_auto_df['Income'].mean()

print(f"Average income for Personal Auto policyholders: {average_income_personal:.2f}")
print(f"Average income for Corporate Auto policyholders: {average_income_corporate:.2f}")

Average income for Personal Auto policyholders: 38180.70
Average income for Corporate Auto policyholders: 41390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [56]:

# Calculating the 75th percentile
percentile_75 = df['Total Claim Amount'].quantile(0.75)

# Calculating the 25th percentile
percentile_25 = df['Total Claim Amount'].quantile(0.25)

# Dataframes for customers with claim amounts greater than the 75th percentile and lower than the 25th percentile
high_claims_df = df[df['Total Claim Amount'] > percentile_75]
low_claims_df = df[df['Total Claim Amount'] < percentile_25]

# Summary statistics for the high and low claim amount customers
summary_stats = high_claims_df.describe()
summary_stats = low_claims_df.describe()

# Summary statistics
print("High Claim Amount Customers:")
print(high_claims_df)
print("Low Claim Amount Customers:")
print(low_claims_df)
print("\nSummary Statistics:")
print(summary_stats)

High Claim Amount Customers:
     Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              899704.02%  54230.0   

      Monthly Premium 